In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [309]:
a = pd.read_excel('recommender_data.xlsx')
a.head()

song id                                          tag count  \
0        0           {'비오는날': 2, '드라이브': 1, '추억': 1, '회상': 1}   
1        3  {'드라이브': 2, '여행': 1, '기분전환': 3, '신나는': 3, '클럽'...   
2        5           {'비오는날': 1, '휴식': 1, '힐링': 1, '기분전환': 1}   
3        6                                 {'산책': 1, '조깅': 1}   
4        9                                          {'위로': 1}   

                        tag max                   emotion count  \
0                      ['비오는날']                      {'우울한': 1}   
1               ['기분전환', '신나는']                      {'행복한': 2}   
2  ['비오는날', '휴식', '힐링', '기분전환']  {'우울한': 1, '잔잔한': 1, '행복한': 2}   
3                  ['산책', '조깅']            {'행복한': 1, '경쾌한': 1}   
4                        ['위로']                      {'우울한': 1}   

      emotion max  
0         ['우울한']  
1         ['행복한']  
2         ['행복한']  
3  ['행복한', '경쾌한']  
4         ['우울한']

In [310]:
a['emotion max'].value_counts()

['행복한']                                159669
['잔잔한']                                 79734
['경쾌한']                                 35646
['우울한']                                 25704
['행복한', '잔잔한']                          15265
                                        ...  
['무서운', '우울한', '행복한', '잔잔한', '경쾌한']         1
['편안한', '잔잔한', '우울한', '무서운']                1
['경쾌한', '행복한', '우울한', '무서운']                1
['울고싶은', '행복한', '잔잔한', '경쾌한']               1
['행복한', '편안한', '무서운', '우울한']                1
Name: emotion max, Length: 758, dtype: int64

In [311]:
sad = []
happy = []
peace = []
light = []
scary = []
comfort = []
cry = []
nervous = []

sad_to_csv = []
happy_to_csv = []
peace_to_csv = []
light_to_csv = []
scary_to_csv = []
comfort_to_csv = []
cry_to_csv = []
nervous_to_csv = []

for i in range(len(a)):
    if '우울한' in a['emotion max'][i]:
        sad.append(a['song id'][i])
        sad_to_csv.append(i)
    if '행복한' in a['emotion max'][i]:
        happy.append(a['song id'][i])
        happy_to_csv.append(i)
    if '잔잔한' in a['emotion max'][i]:
        peace.append(a['song id'][i])
        peace_to_csv.append(i)
    if '경쾌한' in a['emotion max'][i]:
        light.append(a['song id'][i])
        light_to_csv.append(i)
    if '무서운' in a['emotion max'][i]:
        scary.append(a['song id'][i])
        scary_to_csv.append(i)
    if '편안한' in a['emotion max'][i]:
        comfort.append(a['song id'][i])
        comfort_to_csv.append(i)
    if '울고싶은' in a['emotion max'][i]:
        cry.append(a['song id'][i])
        cry_to_csv.append(i)
    if '긴장되는' in a['emotion max'][i]:
        nervous.append(a['song id'][i])
        nervous_to_csv.append(i)

In [312]:
a.iloc[sad_to_csv].to_csv('sad_songs.csv', encoding='utf-8-sig')
a.iloc[happy_to_csv].to_csv('happy_songs.csv', encoding='utf-8-sig')
a.iloc[peace_to_csv].to_csv('peace_songs.csv', encoding='utf-8-sig')
a.iloc[light_to_csv].to_csv('light_songs.csv', encoding='utf-8-sig')
a.iloc[scary_to_csv].to_csv('scary_songs.csv', encoding='utf-8-sig')
a.iloc[comfort_to_csv].to_csv('comfort_songs.csv', encoding='utf-8-sig')
a.iloc[cry_to_csv].to_csv('cry_songs.csv', encoding='utf-8-sig')
a.iloc[nervous_to_csv].to_csv('nervous_songs.csv', encoding='utf-8-sig')

In [314]:
a.iloc[nervous_to_csv]

song id                                          tag count  \
11           18                                        {'스트레스': 1}   
18           34                                        {'스트레스': 1}   
22           40                     {'기분전환': 1, '분위기': 1, '섹시': 1}   
42           81                                        {'스트레스': 1}   
67          127  {'겨울': 1, '스트레스': 1, '출근': 1, '드라이브': 1, '퇴근':...   
...         ...                                                ...   
395226   707729                               {'스트레스': 1, '밴드': 1}   
395250   707774                             {'스트레스': 1, '드라이브': 1}   
395272   707809                      {'추억': 1, '스트레스': 1, '회상': 1}   
395331   707926  {'까페': 1, '분위기': 1, '달콤': 1, '기분전환': 2, '스트레스'...   
395337   707938                                        {'스트레스': 1}   

                                        tag max  \
11                                     ['스트레스']   
18                                     ['스트레스']   
22                        ['기분전환', '분위기', '섹시']   
42                                     ['스트레스']   
67      ['겨울', '스트레스', '출근', '드라이브', '퇴근', '집']   
...                                         ...   
395226                           ['스트레스', '밴드']   
395250                         ['스트레스', '드라이브']   
395272                     ['추억', '스트레스', '회상']   
395331                         ['기분전환', '스트레스']   
395337                                 ['스트레스']   

                                    emotion count             emotion max  
11                                    {'긴장되는': 1}                ['긴장되는']  
18                                    {'긴장되는': 1}                ['긴장되는']  
22                {'행복한': 1, '잔잔한': 1, '긴장되는': 1}  ['행복한', '잔잔한', '긴장되는']  
42                                    {'긴장되는': 1}                ['긴장되는']  
67      {'잔잔한': 1, '긴장되는': 2, '행복한': 1, '편안한': 2}         ['긴장되는', '편안한']  
...                                           ...                     ...  
395226                      {'긴장되는': 1, '경쾌한': 1}         ['긴장되는', '경쾌한']  
395250                      {'긴장되는': 1, '행복한': 1}         ['긴장되는', '행복한']  
395272            {'행복한': 1, '긴장되는': 1, '잔잔한': 1}  ['행복한', '긴장되는', '잔잔한']  
395331                      {'행복한': 1, '긴장되는': 1}         ['행복한', '긴장되는']  
395337                                {'긴장되는': 1}                ['긴장되는']  

[17299 rows x 5 columns]

In [43]:
a['song id'].values

array([     0,      3,      5, ..., 707982, 707986, 707987])

In [44]:
train = pd.read_json('melon_music/train.json', typ = 'frame')
train.rename(columns={'tags' : '태그 리스트','id' : '플레이리스트 ID', 'plylst_title' : '플레이리스트 제목', 'songs' : '곡 리스트', 'like_cnt' : '좋아요 개수', 'updt_date' : '수정 날' }, inplace=True)
train.head()

태그 리스트  플레이리스트 ID  \
0                                                [락]      61281   
1                                           [추억, 회상]      10532   
2                                          [까페, 잔잔한]      76951   
3  [연말, 눈오는날, 캐럴, 분위기, 따듯한, 크리스마스캐럴, 겨울노래, 크리스마스,...     147456   
4                                               [댄스]      27616   

               플레이리스트 제목                                              곡 리스트  \
0                여행같은 음악  [525514, 129701, 383374, 562083, 297861, 13954...   
1                요즘 너 말야  [432406, 675945, 497066, 120377, 389529, 24427...   
2  편하게, 잔잔하게 들을 수 있는 곡.-  [83116, 276692, 166267, 186301, 354465, 256598...   
3  크리스마스 분위기에 흠뻑 취하고 싶을때  [394031, 195524, 540149, 287984, 440773, 10033...   
4               추억의 노래 ㅋ  [159327, 553610, 5130, 645103, 294435, 100657,...   

   좋아요 개수                     수정 날  
0      71  2013-12-19 18:36:19.000  
1       1  2014-12-02 16:19:42.000  
2      17  2017-08-28 07:09:34.000  
3      33  2019-12-05 15:15:18.000  
4       9  2011-10-25 13:54:56.000

In [45]:
def filter_songs(songs):
    filtered_songs = []
    for i in range(len(songs)):
        if songs[i] in a['song id'].values:
            filtered_songs.append(songs[i])
    return filtered_songs

train['filter_songs'] = train['곡 리스트'].apply(lambda x: filter_songs(x))

In [46]:
train.head()

태그 리스트  플레이리스트 ID  \
0                                                [락]      61281   
1                                           [추억, 회상]      10532   
2                                          [까페, 잔잔한]      76951   
3  [연말, 눈오는날, 캐럴, 분위기, 따듯한, 크리스마스캐럴, 겨울노래, 크리스마스,...     147456   
4                                               [댄스]      27616   

               플레이리스트 제목                                              곡 리스트  \
0                여행같은 음악  [525514, 129701, 383374, 562083, 297861, 13954...   
1                요즘 너 말야  [432406, 675945, 497066, 120377, 389529, 24427...   
2  편하게, 잔잔하게 들을 수 있는 곡.-  [83116, 276692, 166267, 186301, 354465, 256598...   
3  크리스마스 분위기에 흠뻑 취하고 싶을때  [394031, 195524, 540149, 287984, 440773, 10033...   
4               추억의 노래 ㅋ  [159327, 553610, 5130, 645103, 294435, 100657,...   

   좋아요 개수                     수정 날  \
0      71  2013-12-19 18:36:19.000   
1       1  2014-12-02 16:19:42.000   
2      17  2017-08-28 07:09:34.000   
3      33  2019-12-05 15:15:18.000   
4       9  2011-10-25 13:54:56.000   

                                        filter_songs  
0  [525514, 129701, 562083, 297861, 139541, 65029...  
1  [432406, 675945, 497066, 120377, 389529, 24427...  
2  [83116, 276692, 166267, 186301, 354465, 256598...  
3  [394031, 195524, 540149, 287984, 440773, 10033...  
4  [159327, 553610, 5130, 645103, 294435, 100657,...

In [47]:
train.shape

(115071, 7)

In [48]:
idx_list = []

for i in range(len(train)):
    if len(train['filter_songs'][i]) >= 5:
        idx_list.append(i)

In [53]:
df = train.iloc[idx_list]
#df.to_csv('genre_filtered_songs.csv', encoding='utf-8-sig')

In [63]:
df = pd.read_csv('genre_filtered_songs.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)
print(df.shape)

(112183, 7)


In [64]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)

In [65]:
print(train_df.shape)
print(test_df.shape)

(89746, 7)
(22437, 7)


In [56]:
import fasttext
import fasttext.util
#fasttext.util.download_model('ko', if_exists='ignore')
ft = fasttext.load_model('../cc.ko.300.bin')

In [67]:
train_df.reset_index(inplace=True, drop=True)
test_df.reset_index(inplace=True, drop=True)

In [69]:
from tqdm import tqdm

# title
train_df['tag_vec_ft'] = '0'

vec = []
for i in tqdm(range(len(train_df))):
    train_df['태그 리스트'][i] = str(train_df['태그 리스트'][i])
    train_df['태그 리스트'][i] = train_df['태그 리스트'][i].replace('[', '').replace(']', '').replace("'", '')
    train_df['tag_vec_ft'][i] = ft.get_sentence_vector(train_df['태그 리스트'][i]).reshape((1,-1))

/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/845809519.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_df['tag_vec_ft'] = '0'
  0%|                                                 | 0/89746 [00:00<?, ?it/s]/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/845809519.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_df['태그 리스트'][i] = str(train_df['태그 리스트'][i])
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: SettingWithCopyWarning: 
A value is tr

In [70]:
from tqdm import tqdm

# title
test_df['tag_vec_ft'] = '0'

vec = []
for i in tqdm(range(len(test_df))):
    test_df['태그 리스트'][i] = str(test_df['태그 리스트'][i])
    test_df['태그 리스트'][i] = test_df['태그 리스트'][i].replace('[', '').replace(']', '').replace("'", '')
    test_df['tag_vec_ft'][i] = ft.get_sentence_vector(test_df['태그 리스트'][i]).reshape((1,-1))

/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_df['tag_vec_ft'] = '0'
  0%|                                                 | 0/22437 [00:00<?, ?it/s]/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_df['태그 리스트'][i] = str(test_df['태그 리스트'][i])
/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fro

In [71]:
#train_df.to_csv('processed_fast_train.csv', encoding='utf-8-sig')
#test_df.to_csv('processed_fast_test.csv', encoding='utf-8-sig')

In [76]:
print(len(test_df))
new_test_df = test_df.head(100)
print(len(new_test_df))

22437
100


In [77]:
## 추천 구하기
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics.pairwise import cosine_similarity

rec_result = []

for j in tqdm(range(len(new_test_df))):
    target = new_test_df['tag_vec_ft'].values[j][0].reshape(1, -1)
    
    train_df['cos_sim'] = train_df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, target))

    rec_result.append((train_df.iloc[np.argmax(train_df['cos_sim'])]['곡 리스트'],
                      np.argmax(train_df['cos_sim']), max(train_df['cos_sim'])))
    

100%|█████████████████████████████████████████| 100/100 [46:06<00:00, 27.67s/it]


In [78]:
# 평가하기
class Evaluator:
    def _idcg(self, l):
        return sum((1.0 / np.log(i + 2) for i in range(l)))

    def __init__(self):
        self._idcgs = [self._idcg(i) for i in range(101)]

    def _ndcg(self, gt, rec):
        dcg = 0.0
        for i, r in enumerate(rec):
            if r in gt:
                dcg += 1.0 / np.log(i + 2)
        #print(len(gt))
        #print(len(rec))
        #print()
        return dcg / self._idcgs[len(gt)]
    
evaluator = Evaluator()

In [95]:
ndcg_total_score = []

for i in tqdm(range(len(new_test_df))):
    num = i

    true_relevance = new_test_df['곡 리스트'][num][:100]
    scores = rec_result[num][0][:100]
            
    ndcg_total_score.append(evaluator._ndcg(true_relevance, scores))    
print(np.mean(ndcg_total_score))

100%|███████████████████████████████████████| 100/100 [00:00<00:00, 3176.32it/s]

1.003245176529229


In [148]:
#### 무시 ~~~~~
ndcg_total_score = []

for i in tqdm(range(len(new_test_df))):
    num = i

    true_relevance = new_test_df['곡 리스트'][num].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')[:100]
    scores = rec_result[num][0].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')[:100]
            
    ndcg_total_score.append(evaluator._ndcg(true_relevance, scores))    
print(np.mean(ndcg_total_score))

  1%|▍                                         | 1/100 [00:00<00:00, 941.69it/s]


IndexError: list index out of range

#### 결론: # pre-trained fasttext를 이용한 임베딩시, test set 100개의 ndcg 평균: 1.003

## 유저 관점 테스트

In [321]:
song = pd.read_json('melon_music/song_meta.json',typ = 'frame')
song.rename(columns={"song_gn_dtl_gnr_basket":"곡 세부 장르 리스트", "issue_date":"발매일", "album_name": "앨범 명", "album_id":"앨범 ID", "artist_id_basket":"아티스트 ID 리스트", "song_name":"곡 제목", "song_gn_gnr_basket":"곡 장르 리스트", "artist_name_basket":"아티스트 리스트","id":"곡 ID"},inplace= True )
song.head()

곡 세부 장르 리스트       발매일  \
0          [GN0901]  20140512   
1  [GN1601, GN1606]  20080421   
2          [GN0901]  20180518   
3  [GN1102, GN1101]  20151016   
4  [GN1802, GN1801]  20110824   

                                                앨범 명    앨범 ID 아티스트 ID 리스트  \
0                       불후의 명곡 - 7080 추억의 얄개시대 팝송베스트  2255639      [2727]   
1                      Bach : Partitas Nos. 2, 3 & 4   376431     [29966]   
2                                                Hit  4698747      [3361]   
3  Feeling Right (Everything Is Nice) (Feat. Popc...  2644882    [838543]   
4                                            그남자 그여자  2008470    [560160]   

                                                곡 제목  곡 장르 리스트  \
0                                           Feelings  [GN0900]   
1  Bach : Partita No. 4 In D Major, BWV 828 - II....  [GN1600]   
2                    Solsbury Hill (Remastered 2002)  [GN0900]   
3  Feeling Right (Everything Is Nice) (Feat. Popc...  [GN1100]   
4                                            그남자 그여자  [GN1800]   

            아티스트 리스트  곡 ID  
0  [Various Artists]     0  
1   [Murray Perahia]     1  
2    [Peter Gabriel]     2  
3           [Matoma]     3  
4         [Jude Law]     4

In [322]:
songs_with_tags = pd.merge(song, a, left_on='곡 ID', right_on = 'song id', how='outer')#['song id']
songs_with_tags = songs_with_tags.set_index('곡 ID', drop=True)

In [235]:
# 유저 관점:

user_tag = '힐링, 쉼, 휴식, 고요'
emotion_result = ('잔잔한', 0.7), ('편안한', 0.3)
top_n = 5
max_song_list = 100
################

emotion_n = len(emotion_result)

sa_active = 0
h_active = 0
p_active = 0
l_active = 0
sc_active = 0
co_active = 0
cr_active = 0
n_active = 0

for emo_idx in range(len(emotion_result)):
    emo = emotion_result[emo_idx][0]
    if emo == '잔잔한':
        p_active = emotion_result[emo_idx][1]
    if emo == '편안한':
        co_active = emotion_result[emo_idx][1]
    if emo == '우울한':
        sa_active = emotion_result[emo_idx][1]
    if emo == '행복한':
        h_active = emotion_result[emo_idx][1]
    if emo == '경쾌한':
        l_active = emotion_result[emo_idx][1]
    if emo == '무서운':
        sc_active = emotion_result[emo_idx][1]
    if emo == '울고싶은':
        cr_active = emotion_result[emo_idx][1]
    if emo == '긴장되는':
        n_active = emotion_result[emo_idx][1]

In [236]:
user_tag_vectors = ft.get_sentence_vector(user_tag).reshape((1,-1))

## 추천 구하기
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics.pairwise import cosine_similarity

rec_result = []
    
train_df['cos_sim'] = train_df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, user_tag_vectors))
## 시간 오래걸리는데 train 사이즈를 줄일까?

In [238]:
a[a['song id'] == 96310]


song id                                          tag count tag max  \
53528    96310  {'햇살': 4, '봄': 49, '카페': 80, '봄바람': 6, '주말': 2...  ['사랑']   

      emotion count emotion max  
53528    {'행복한': 1}     ['행복한']

In [316]:
import heapq
fr = train_df['cos_sim']
n_result = heapq.nlargest(max_song_list, range(len(fr)), key=fr.__getitem__)

print('active 값:', sa_active, h_active, p_active, l_active, sc_active, co_active, cr_active, n_active)

found_list =[]
found = 0
flag = 1

for i in range(max_song_list):
    sc_list = []
    rec_result = []
    rec_result.append((train_df.iloc[n_result[i]]['곡 리스트'],
                   train_df.iloc[n_result[i]]['cos_sim']))

    user_rec_songs = rec_result[0][0].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')#[:100]
    user_rec_songs = list(map(int, user_rec_songs))
    print('similarity', rec_result[0][1])

    for song in user_rec_songs:
        sc = 0
        if sa_active:
            if song in sad:
                sc += sa_active
        if h_active:
            if song in happy:
                sc += h_active
        if p_active:
            if song in peace:
                sc += p_active
        if l_active:
            if song in light:
                sc += l_active
        if sc_active:
            if song in scary:
                sc += sc_active
        if co_active:
            if song in comfort:
                sc += co_active
        if cr_active:
            if song in cry:
                sc += cr_active
        if n_active:
            if song in nerbous:
                sc += n_active
        
        if sc>0:
            print('song:', song, 'sc:', sc)
            sc_list.append(sc)

            found += 1
            found_list.append(song)
        
    if found >= top_n:
        break


active 값: 0 0 0.7 0 0 0.3 0 0
similarity [[0.83917713]]
song: 604870 sc: 0.7
song: 480675 sc: 0.7
song: 698648 sc: 0.7
song: 125822 sc: 0.7
song: 431225 sc: 0.7


In [317]:
# 후보
found_list

[604870, 480675, 698648, 125822, 431225]

In [318]:
# sc 높은 순으로 추천
song_result = heapq.nlargest(top_n, range(len(sc_list)), key=sc_list.__getitem__)
fin_song_result = [found_list[song_result[i]] for i in range(top_n)]
fin_song_result

[604870, 480675, 698648, 125822, 431225]

In [324]:
song.iloc[fin_song_result]

곡 세부 장르 리스트       발매일  \
604870           [GN2502, GN0105, GN2501, GN0101, GN2505]  20151012   
480675                   [GN2502, GN0601, GN2501, GN0606]  20120710   
698648  [GN0105, GN0101, GN2505, GN2502, GN1501, GN250...  20161115   
125822           [GN0105, GN2505, GN2501, GN0101, GN2503]  20161123   
431225                                   [GN0105, GN0101]  20160911   

                                       앨범 명     앨범 ID  \
604870                               I Mean   2644202   
480675                              Goodbye   2136832   
698648                캐리어를 끄는 여자 OST Part.6  10015096   
125822  Jelly box 꽃길 (Prod. By 지코(ZICO)) 세정  10016726   
431225                             복면가왕 76회   2710626   

                             아티스트 ID 리스트                         곡 제목  \
604870  [672287, 672288, 672289, 672296]                       여기 있을게   
480675                          [623451]        Goodbye (With Matoki)   
698648                          [670611]                        인형의 꿈   
125822                          [968261]           꽃길 (Prod. By ZICO)   
431225                          [672287]  바람이 분다 (오늘 밤엔 어둠이 무서워요 석봉이)   

                        곡 장르 리스트                          아티스트 리스트    곡 ID  
604870          [GN2500, GN0100]  [서은광, 이창섭, 임현식 (비투비), 육성재 (비투비)]  604870  
480675          [GN2500, GN0600]                           [B.A.P]  480675  
698648  [GN2500, GN1500, GN0100]                        [태일 (블락비)]  698648  
125822          [GN2500, GN0100]                        [세정 (구구단)]  125822  
431225                  [GN0100]                             [서은광]  431225

In [325]:
songs_with_tags.iloc[fin_song_result]

곡 세부 장르 리스트       발매일  \
곡 ID                                                                  
604870           [GN2502, GN0105, GN2501, GN0101, GN2505]  20151012   
480675                   [GN2502, GN0601, GN2501, GN0606]  20120710   
698648  [GN0105, GN0101, GN2505, GN2502, GN1501, GN250...  20161115   
125822           [GN0105, GN2505, GN2501, GN0101, GN2503]  20161123   
431225                                   [GN0105, GN0101]  20160911   

                                       앨범 명     앨범 ID  \
곡 ID                                                    
604870                               I Mean   2644202   
480675                              Goodbye   2136832   
698648                캐리어를 끄는 여자 OST Part.6  10015096   
125822  Jelly box 꽃길 (Prod. By 지코(ZICO)) 세정  10016726   
431225                             복면가왕 76회   2710626   

                             아티스트 ID 리스트                         곡 제목  \
곡 ID                                                                    
604870  [672287, 672288, 672289, 672296]                       여기 있을게   
480675                          [623451]        Goodbye (With Matoki)   
698648                          [670611]                        인형의 꿈   
125822                          [968261]           꽃길 (Prod. By ZICO)   
431225                          [672287]  바람이 분다 (오늘 밤엔 어둠이 무서워요 석봉이)   

                        곡 장르 리스트                          아티스트 리스트   song id  \
곡 ID                                                                           
604870          [GN2500, GN0100]  [서은광, 이창섭, 임현식 (비투비), 육성재 (비투비)]  604870.0   
480675          [GN2500, GN0600]                           [B.A.P]  480675.0   
698648  [GN2500, GN1500, GN0100]                        [태일 (블락비)]  698648.0   
125822          [GN2500, GN0100]                        [세정 (구구단)]  125822.0   
431225                  [GN0100]                             [서은광]  431225.0   

                                                tag count        tag max  \
곡 ID                                                                       
604870  {'따뜻한': 1, '사랑': 11, '이별': 6, '까페': 1, '힐링': 1...         ['휴식']   
480675  {'지친하루': 1, '위로': 5, '휴식': 12, '힐링': 11, '신나는'...         ['휴식']   
698648  {'힐링': 2, '휴식': 4, '봄': 2, '달달': 2, '커플': 1, '...         ['휴식']   
125822  {'후회': 4, '눈물': 19, '사랑': 75, '새벽': 60, '우울': ...  ['힐링', '잔잔한']   
431225  {'사랑': 2, '이별': 2, '댄스': 3, '휴식': 3, '눈오는날': 1...   ['댄스', '휴식']   

               emotion count     emotion max  
곡 ID                                          
604870            {'잔잔한': 1}         ['잔잔한']  
480675            {'잔잔한': 1}         ['잔잔한']  
698648            {'잔잔한': 1}         ['잔잔한']  
125822  {'행복한': 1, '잔잔한': 1}  ['행복한', '잔잔한']  
431225  {'경쾌한': 1, '잔잔한': 1}  ['경쾌한', '잔잔한']

In [303]:
355771 in peace

True